### PR TITLE
Use imageStore.Children instead of whole the Map

### PR DESCRIPTION
**Pull request by cpuguy83**
daemon cache was getting the whole image map and then iterating through
it to find children. This information is already stored in the image
store.

Prior to this change building the docker repo with a full cache took 30
seconds.
After it takes between 15 seconds or less (As low as 9 seconds).
This is an improvement on docker 1.9.1 which hovered around 17 seconds.

### CODE REVIEW

**Comment by cpuguy83** [comment]
I also tried to cache `*Image` references in the image store instead of reading from JSON every time, but for some reason over (a short) time the image comparison would fail, and that was just with running the build a few times and nothing else.

**Comment by cpuguy83** [comment]
Tentatively adding to the 1.10 milestone since 1.10 is significantly slower.

**Comment by cpuguy83** [comment]
Also I think the cache probably won't help much. While I was able to consistently get 9-10s for building our Dockerfile with it, with just the change in the PR it is frequently around 10s anyway (though less consistently).

**Comment by tonistiigi** [comment]
LGTM, maybe also remove `daemon.Map()`

**Comment by cpuguy83** [comment]
Updated.
For the `FROM scratch` case, the lookup for the first step is still slow, but everything after will be fast.

File: daemon/daemon.go [hunk 1]
```diff
@@ -1295,30 +1295,39 @@ func (daemon *Daemon) GetRemappedUIDGID() (int, int) {
...
+			if err != nil {
+				return nil, fmt.Errorf("unable to find image %q", id)
+			}
 
-	var siblings []image.ID
-	for id, img := range imgs {
-		if img.Parent == imgID {
-			siblings = append(siblings, id)
+			if runconfig.Compare(&img.ContainerConfig, config) {
+				if match == nil || match.Created.Before(img.Created) {
```
Comment (LK4D4) [inline daemon/daemon.go]: Coul you write comment for this `Created.Before` logic please? It's not very obvious why we need it.

File: daemon/daemon.go [hunk 2]
```diff
@@ -1295,30 +1295,39 @@ func (daemon *Daemon) GetRemappedUIDGID() (int, int) {
...
-	for _, id := range siblings {
-		img, ok := imgs[id]
-		if !ok {
-			return nil, fmt.Errorf("unable to find image %q", id)
-		}
-		if runconfig.Compare(&img.ContainerConfig, config) {
-			if match == nil || match.Created.Before(img.Created) {
-				match = img
+	// In this case, this is `FROM scratch`, which isn't an actual image.
+	if imgID == image.ID("") {
```
Comment (tonistiigi) [inline daemon/daemon.go]: Cast is not needed when comparing with empty.

File: daemon/daemon.go (hunk 1 above)
Comment (cpuguy83) [inline daemon/daemon.go]: Actually it doesn't make all that much sense to me.
Based on the comment on the function, we want the earliest match, but here we're setting the match to `img` which is created before `match`? Logic error or am I reading incorrectly?

Comment (tonistiigi) [inline daemon/daemon.go]: I think we want the most up-to-date match. So if the previous match is older we replace it.

Comment (LK4D4) [inline daemon/daemon.go]: I think @tonistiigi is right. We're looking for latest. Still deserves comment, because even author was confused :)

Comment (cpuguy83) [inline daemon/daemon.go]: I didn't author this :) but yes, adding a comment and fixing the docstring for this fn which is ambiguous on the matter.

**Comment by cpuguy83** [comment]
updated

**Comment by tonistiigi** [comment]
LGTM

**Comment by tiborvass** [comment]
LGTM

**Comment by cpuguy83** [comment]
gccgo: `21:00:52 Finished: SUCCESS`, but didn't notify back to github.
All green + 2 LGTM's, merging.

**Pull request merged**